rclcpp_examples/src/topics/allocator_example.cpp

```diff
@@ -0,0 +1,194 @@
+  rclcpp::init(argc, argv);
+
+  auto node = rclcpp::Node::make_shared("allocator_example", false);
+  // To use intra-process, comment out the above line and uncomment the following:
+  /*
+  auto context = rclcpp::contexts::default_context::get_global_default_context();
+  auto ipm_state =
+    std::make_shared<rclcpp::intra_process_manager::IntraProcessManagerState<MyAllocator<>>>();
+  // Constructs the intra-process manager with a custom allocator.
+  context->get_sub_context<rclcpp::intra_process_manager::IntraProcessManager>(ipm_state);
+  auto node = rclcpp::Node::make_shared("allocator_example", true);
+  */
+
+  uint32_t counter = 0;
+  auto callback = [&counter](std_msgs::msg::UInt32::SharedPtr msg) -> void
+    {
+      (void)msg;
+      counter++;
+    };
+
+  // Create a custom allocator and pass the allocator to the publisher and subscriber.
+  auto alloc = std::make_shared<MyAllocator<void>>();
+  auto publisher = node->create_publisher<std_msgs::msg::UInt32>("allocator_example", 10, alloc);
+  auto msg_mem_strat =
+    std::make_shared<rclcpp::message_memory_strategy::MessageMemoryStrategy<std_msgs::msg::UInt32,
+    MyAllocator<>>>(alloc);
+  auto subscriber = node->create_subscription<std_msgs::msg::UInt32>(
+    "allocator_example", 10, callback, nullptr, false, msg_mem_strat, alloc);
+
+  // Create a MemoryStrategy, which handles the allocations made by the Executor during the
+  // execution path, and inject the MemoryStrategy into the Executor.
+  std::shared_ptr<rclcpp::memory_strategy::MemoryStrategy> memory_strategy =
+    std::make_shared<AllocatorMemoryStrategy<MyAllocator<>>>(alloc);
+  rclcpp::executors::SingleThreadedExecutor executor(memory_strategy);
+
+  // Add our node to the executor.
+  executor.add_node(node);
+
+  // Create a message with the custom allocator, so that when the Executor deallocates the
+  // message on the execution path, it will use the custom deallocate.
+  auto msg = std::allocate_shared<std_msgs::msg::UInt32>(*alloc.get());
+
+  rclcpp::utilities::sleep_for(std::chrono::milliseconds(1));
+  is_running = true;
+
+  uint32_t i = 0;
+  while (rclcpp::ok()) {
+    msg->data = i;
+    i++;
+    publisher->publish(msg);
+    rclcpp::utilities::sleep_for(std::chrono::milliseconds(1));
+    executor.spin_some();
+  }
+  is_running = false;
+
+  uint32_t final_global_allocs = global_runtime_allocs;
+  uint32_t final_global_deallocs = global_runtime_deallocs;
+  std::cout << "Global new was called " << final_global_allocs
+            << " times during spin" << std::endl;
+  std::cout << "Global delete was called " << final_global_deallocs
+            << " times during spin" << std::endl;
+
+  std::cout << "Allocator new was called " << num_allocs
+            << " times during spin" << std::endl;
+  std::cout << "Allocator delete was called " << num_deallocs
+            << " times during spin" << std::endl;
+  rclcpp::utilities::sleep_for(std::chrono::milliseconds(1));
```

Comment:
Sometimes I see the program exit before these statements print to screen. Maybe when I change them to printfs this problem will be fixed.